blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+
+Old value = (void *) 0x7f0490b784d0 <jl_system_image_data+16912>
+New value = (void *) 0x0
+jl_svecset (x=0x7f0490b784d0 <jl_system_image_data+16912>, i=<optimized out>, t=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/julia.h:840
+840 /buildworker/worker/package_linux64/build/src/julia.h: No such file or directory.
+(rr) bt
+#0  jl_svecset (x=0x7f0490b784d0 <jl_system_image_data+16912>, i=<optimized out>, t=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/julia.h:840
+#1  jl_f_tuple (F=<optimized out>, args=0x42f54590, nargs=46200)
+    at /buildworker/worker/package_linux64/build/src/builtins.c:728
+```
+
+For reference (from the 1.5 release branch):
+
+```
+JL_CALLABLE(jl_f_tuple)
+{
+    size_t i;
+    if (nargs == 0)
+        return (jl_value_t*)jl_emptytuple;
+    jl_datatype_t *tt;
+    if (nargs < jl_page_size / sizeof(jl_value_t*)) {
+        jl_value_t **types = (jl_value_t**)alloca(nargs * sizeof(jl_value_t*));
+        for (i = 0; i < nargs; i++)
+            types[i] = jl_typeof(args[i]);
+        tt = jl_inst_concrete_tupletype_v(types, nargs);
+    }
+    else {
+        jl_svec_t *types = jl_alloc_svec_uninit(nargs);
+        JL_GC_PUSH1(&types);
+        for (i = 0; i < nargs; i++)
+            jl_svecset(types, i, jl_typeof(args[i]));
+        tt = jl_inst_concrete_tupletype(types);
+        JL_GC_POP();
+    }
+    if (tt->instance != NULL)
+        return tt->instance;
+    jl_ptls_t ptls = jl_get_ptls_states();
+    jl_value_t *jv = jl_gc_alloc(ptls, jl_datatype_size(tt), tt);
+    for (i = 0; i < nargs; i++)
+        set_nth_field(tt, (void*)jv, i, args[i]);
+    return jv;
+}
+```
+
+So we're just getting the type pointer from some array. We can find out which one:
+
+```
+(rr) disas
+   0x00007f049e2d8530 <+240>: mov    -0x8(%r12,%r13,1),%rax
+   0x00007f049e2d8535 <+245>: mov    -0x8(%rax),%rax
+   0x00007f049e2d8539 <+249>: and    $0xfffffffffffffff0,%rax
+   0x00007f049e2d853d <+253>: test   %rax,%rax
+   0x00007f049e2d8540 <+256>: mov    %rax,(%r14,%r13,1)
+=> 0x00007f049e2d8544 <+260>: je     0x7f049e2d8553 <jl_f_tuple+275>
+(rr) rsi
+(rr) rsi
+(rr) rsi
+(rr) rsi
+=> 0x00007f049e2d8535 <+245>: mov    -0x8(%rax),%rax
+   0x00007f049e2d8539 <+249>: and    $0xfffffffffffffff0,%rax
+   0x00007f049e2d853d <+253>: test   %rax,%rax
+   0x00007f049e2d8540 <+256>: mov    %rax,(%r14,%r13,1)
+   0x00007f049e2d8544 <+260>: je     0x7f049e2d8553 <jl_f_tuple+275>
+(rr) p /x $rax
+$5 = 0x7f0478f5aa10
+(rr) p jl_($rax)
+Array{Any, (2,)}[
+  Array{Float64, (1,)}[43.1183],
+  0.86978]
+$6 = void
+```
+
+We can also watch this array's type pointer to see if there's anything
+unexpected that messes with it:
+
+```
+(rr) p &((uint64_t**)0x7f0478f5aa10)[-1]
+$9 = (void **) 0x7f0478f5aa08
+(rr) p jl_((uint64_t)*$9 & ~0x3)
+Array{Any, 1}
+```
+
+(note that we masked out the low bits, which julia does use as mentioned above - and
```

Comment:
```suggestion
(note that we masked out the low bits, which Julia does use as mentioned above - and
```

Reply:
This is an implementation detail, so the style guide suggests julia (the program/implementation) rather than Julia (the abstract language). I'll code-format it.